Add CBOR Support to libyang  (#2130)

MeherRushi opened the pull request:
## Summary                                                                                  
  Adds CBOR (Concise Binary Object Representation) support to libyang, addressing #2130. This
  enables parsing, validating, and printing instance data in CBOR format per [RFC 9254](https://datatracker.ietf.org/doc/rfc9254/), using [[libcbor](https://github.com/PJK/libcbor)](https://github.com/PJK/libcbor) as the underlying library. CBOR support is **optional** — if libcbor is found at build time, support is compiled in automatically; otherwise it is silently skipped.
  ## Changes
  ### New files
  - `CMakeModules/FindCBOR.cmake` — CMake find module for libcbor dependency detection
  - `src/lcbor.c` / `src/lcbor.h` — libcbor abstraction layer (context management, type
  helpers)
  - `src/parser_cbor.c` — CBOR parser: constructs libyang data trees from CBOR input
  - `src/printer_cbor.c` — CBOR printer: serializes libyang data trees to CBOR output
  ### Modified files
  - `CMakeLists.txt` — conditionally compile CBOR sources when libcbor is found
  - `src/parser_internal.h` — add `lycbor_ctx` to parser context union
  - `src/tree_data.c` — CBOR format handling in data tree operations
  - `src/printer_cbor.c` — migrated to current upstream API
  ### New format type
  - `LY_VALUE_CBOR` added to the value format enum
  ## Supported functionality
  - All YANG node types: leaf, leaf-list, list, container, anydata/anyxml, choice/case
  - Metadata parsing and printing
  - Opaque (unknown) nodes
  - RPC/action/notification/reply data trees
  - Named CBOR encoding (JSON-like keys)
  ## PR review feedback addressed
  - **Removed `ENABLE_CBOR_SUPPORT` option** — libcbor is auto-detected via `FindCBOR.cmake`
  (no pkg-config dependency)
  - **Removed file-level `#ifdef` guards** — CMake handles conditional compilation
  - **Removed `LYD_ANYDATA_CBOR`** — uses `LYD_ANYDATA_DATATREE` as recommended
  - **Removed all debug `printf()` statements** and leftover artifacts
  - **Reverted unintended `parser_json.c` changes** — restored to upstream
  - Added `lyd_parse_data_mem_len()` high-level function for parsing CBOR from memory (required because CBOR binary data may contain null bytes, making `strlen()` unreliable)
   - Implemented consistent function structure mirroring JSON counterparts for parsing (name parsing, prefix handling, schema node lookup, metadata, opaque nodes, etc.)
   - Added proper error handling, null handling, array processing, and metadata support
  - **Migrated `parser_cbor.c` to current upstream API:**
    - `lys_find_child_node` instead of removed schema lookup functions
    - Updated `LOGVAL` calls with `lnode` parameter
    - Fixed `LOG_LOCSET`/`LOG_LOCBACK` to single-arg macros
    - Updated `lyd_parser_create_term` and `lyd_parser_create_meta` signatures
    - Updated `lyd_create_any` to new 7-arg signature
    - Removed `lyplg_ext_insert` (removed upstream)
    - Fixed `lyd_node_any` access (`value.tree` → `child`)
    - `LYD_PRINT_WITHSIBLINGS` → `LYD_PRINT_SIBLINGS`
  ## Test plan
  - [x] `cmake` configuration succeeds with libcbor installed
  - [x] Full build completes without errors 
  - [x] Tested on Linux (build + runtime verification)
  ## Future work
  - CBOR SID (Schema Item iDentifier) encoding support
  - Unit tests (to be added in follow-up)
  - Integration with `yanglint` CLI tool

Attempt to resolve #2130

### Key Features
- **Optional Dependency**: CBOR support is controlled via the compile-time flag `ENABLE_CBOR_SUPPORT`, allowing builds with or without CBOR capabilities
- **Full Parsing Support**: Complete CBOR parser that constructs libyang data trees and links them with schema trees for validation
- **Full Printing Support**: CBOR printer with parallel structure to `printer_json.c`, supporting all YANG node types (leaf, leaf-list, list, container, anydata, anyxml, RPC, action, notification)
- **Consistent API**: Mirrors existing JSON functionality for ease of use and maintenance


### TODO 
- [ ] **CBOR SID support** - Currently only named identifiers are supported; add SID (Schema Item iDentifier) support
- [ ] **Finalize input API** - Determine optimal API design for CBOR input handling
- [ ] **Handle missing switch cases** - Complete remaining switch statement cases - some warnings in cmake process 
- [ ] **RPC/anydata edge cases** - Address remaining edge cases for RPC and anydata handling

### Dependencies
- **libcbor** 
- same build processf
---

# Testing :
Considered an `example.json` file : 
```
{
  "ietf-interfaces:interfaces": {
    "interface": [
      {
        "name": "eth0",
        "description": "Loopback",
        "type": "iana-if-type:ethernetCsmacd",
        "enabled": true,
        "oper-status": "up",
        "statistics": {
          "discontinuity-time": "2024-01-01T00:00:00Z"
        }
      }
    ]
  }
}
```

Converted the same into cbor notation and output in `example.cbor` using standard cbor libraries in python (below code - json_to_cbor.py):
```
import json
import cbor2
import time

# Read the JSON file
with open("example.json", "r", encoding="utf-8") as f:
    json_data = json.load(f)

# Write CBOR-encoded output
with open("example.cbor", "wb") as f:
    cbor2.dump(json_data, f)

print("✅ example.cbor written from example.json")

time.sleep(5)

with open("example.cbor", "rb") as f:
    obj = cbor2.load(f)

print(json.dumps(obj, indent=2))
```

Now ran the following json_test.c on the json file with the necessary yang modules `iana-if-type.yang` and `ietf-interfaces.yang` in the directory to see functionality on the json data
```#include <stdio.h>
#include <stdlib.h>
#include <libyang/libyang.h>

int main(void) {
    struct ly_ctx *ctx = NULL;
    struct lyd_node *data = NULL;

    // Context: search current dir only, no auto-loading
    if (ly_ctx_new(".", LY_CTX_NO_YANGLIBRARY, &ctx) != LY_SUCCESS) {
        fprintf(stderr, "Context creation failed.\n");
        return 1;
    }

    // Load only required official modules
    if (lys_parse_path(ctx, "iana-if-type.yang", LYS_IN_YANG, NULL) != LY_SUCCESS ||
        lys_parse_path(ctx, "ietf-interfaces.yang", LYS_IN_YANG, NULL) != LY_SUCCESS) {
        fprintf(stderr, "Module loading failed.\n");
        ly_ctx_destroy(ctx);
        return 1;
    }

    // Parse data
    if (lyd_parse_data_path(ctx, "example.json", LYD_JSON, LYD_PARSE_STRICT, 0, &data) != LY_SUCCESS) {
        fprintf(stderr, "JSON parsing failed.\n");
        ly_ctx_destroy(ctx);
        return 1;
    }

    // Print parsed data
    lyd_print_file(stdout, data, LYD_JSON, 0);

    // Cleanup
    lyd_free_all(data);
    ly_ctx_destroy(ctx);
    return 0;
}
```
Repeated the same with the `cbor_test.c` file (in the same directory) utilizing the cbor features introduced to test the newly built library

```#include <stdio.h>
#include <stdlib.h>
#include <libyang/libyang.h>

int main(void) {
    struct ly_ctx *ctx = NULL;
    struct lyd_node *data = NULL;

    // Context: search current dir only, no auto-loading
    if (ly_ctx_new(".", LY_CTX_NO_YANGLIBRARY, &ctx) != LY_SUCCESS) {
        fprintf(stderr, "Context creation failed.\n");
        return 1;
    }

    // Load only required official modules
    if (lys_parse_path(ctx, "iana-if-type.yang", LYS_IN_YANG, NULL) != LY_SUCCESS ||
        lys_parse_path(ctx, "ietf-interfaces.yang", LYS_IN_YANG, NULL) != LY_SUCCESS) {
        fprintf(stderr, "Module loading failed.\n");
        ly_ctx_destroy(ctx);
        return 1;
    }

    // Parse data
    if (lyd_parse_data_path(ctx, "example.cbor", LYD_CBOR, LYD_PARSE_STRICT, 0, &data) != LY_SUCCESS) {
        fprintf(stderr, "CBOR parsing failed.\n");
        ly_ctx_destroy(ctx);
        return 1;
    }

    // Print parsed to file named "output.cbor"
    // output.cbor should be a file pointer opened in binary write mode
    FILE *output_file = fopen("output.cbor", "wb");
    if (!output_file) {
        fprintf(stderr, "Failed to open output.cbor for writing.\n");
        ly_ctx_destroy(ctx);
        return 1;
    }
    lyd_print_file(output_file, data, LYD_CBOR, 0);
    fclose(output_file);
    // lyd_print_file(stdout, data, LYD_JSON, 0);


    // Cleanup
    lyd_free_all(data);
    ly_ctx_destroy(ctx);
    return 0;
}
```

Following output:
```
$ gcc cbor_test.c -lyang -o cbor
$./cbor 
DEBUG: Parsed CBOR data:
{"ietf-interfaces:interfaces": {"interface": [{"name": "eth0", "description": "Loopback", "type": "iana-if-type:ethernetCsmacd", "enabled": true, "oper-status": "up", "statistics": {"discontinuity-time": "2024-01-01T00:00:00Z"}}]}}
``` 

And we can change the input file from `example.cbor` to `output.cbor` to see the parsed data being parsed properly
```
$ gcc cbor_test.c -lyang -o cbor
(base) meherrushi@MeherRushi:~/Meher/learning/c/cbor/test$ ./cbor 
DEBUG: Parsed CBOR data:
{"ietf-interfaces:interfaces": {"interface": [{"name": "eth0", "description": "Loopback", "type": "iana-if-type:ethernetCsmacd", "enabled": true, "oper-status": "up", "statistics": {"discontinuity-time": "2024-01-01T00:00:00+00:00"}}]}}
```

> note: shall upload the test files on another repo

MeherRushi committed:
feat: add initial CBOR format support with libcbor integration

MeherRushi committed:
refactor: restructure CBOR parser to match libyang conventions

MeherRushi committed:
feat: add CBOR printer skeleton and rename parser functions

MeherRushi committed:
refactor: align CBOR implementation with project structure

MeherRushi committed:
feat: complete CBOR parser with metadata, opaque nodes, and LY_VALUE_CBOR

MeherRushi committed:
feat: implement CBOR printer functionality

michalvasko commented:
Thanks a lot for this effort and I will try to provide some feedback today or tomorrow. But I can immediately tell you to change the *libyang* base branch to `devel`, PRs are never merged into `master` directly.

michalvasko commented on CMakeLists.txt:
```diff
@@ -249,6 +250,7 @@ option(ENABLE_YANGLINT_INTERACTIVE "Enable interactive CLI yanglint" ON)
 option(ENABLE_TOOLS "Build binary tools 'yanglint' and 'yangre'" ON)
 option(ENABLE_COMMON_TARGETS "Define common custom target names such as 'doc' or 'uninstall', may cause conflicts when using add_subdirectory() to build this project" ON)
 option(BUILD_SHARED_LIBS "By default, shared libs are enabled. Turn off for a static build." ON)
+option(ENABLE_CBOR_SUPPORT "Enable CBOR support with libcbor" ON)
```
Actually, an `option` is probably redundant. You can just search for `libcbor` and if found, the support will be added, otherwise not.

michalvasko commented on CMakeLists.txt:
```diff
@@ -316,6 +318,42 @@ if(ENABLE_COVERAGE)
 endif()
 
+if(ENABLE_CBOR_SUPPORT)
+    find_package(PkgConfig)
```
Is `pkg-config` really required? We have used it only when necessary, the `else()` branch should be fine except may be better to add a new `FindCBOR.cmake` module file.

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,131 @@
+ *     https://opensource.org/licenses/BSD-3-Clause
+ */
+
+#ifdef ENABLE_CBOR_SUPPORT
```
Seems redundant if the define is around the whole file, the file should just not be compiled at all if disabled. In all the new files.

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,131 @@
+/**
+ * @file lcbor.h
```
Just `cbor.c` would probably fit better, like `lyb.c` and `lyb.h`.

MeherRushi replied:
The issue is that libcbor library files are also cbor.h - it is causing some build issues - so for simplicity - decided to use lcbor.h/c

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,131 @@
+ */
+void lycbor_ctx_free(struct lycbor_ctx *cborctx)
+{
+    if (cborctx)
```
Redundant `if` and maybe the whole function.

michalvasko commented on src/parser_cbor.c:
```diff
@@ -0,0 +1,2089 @@
+#include "validation.h"
+
+#include <stdio.h>
+#include <string.h>
```
Keep standard includes together and without duplicates.

michalvasko commented on src/parser_cbor.c:
```diff
@@ -0,0 +1,2089 @@
+{
+    if (cbor_isa_uint(item))
+    {
+        printf("%lu", cbor_get_uint64(item));
```
These printf formats tend to throw warnings on some platforms, better to use `PRIu64` and similar.

michalvasko commented on src/parser_cbor.c:
```diff
@@ -0,0 +1,2089 @@
+            if (!node->schema)
+            {
+                /* opaq node */
+                if (strcmp(&meta_container->name.name[1], ((struct lyd_node_opaq *)node)->name.name))
```
A hint, you can use `LYD_NAME(node)` instead of `((struct lyd_node_opaq *)node)->name.name` and it works for both data and opaque nodes.

michalvasko replied:
Not updated, not sure why.

michalvasko commented on src/parser_internal.h:
```diff
@@ -179,6 +179,35 @@ struct lyd_lyb_ctx {
     struct lylyb_ctx *lybctx;      /* LYB context */
 };
 
+#ifdef ENABLE_CBOR_SUPPORT
```
Not really needed, I think, if you just declare `struct lycbor_ctx;`.

michalvasko commented on src/parser_internal.h:
```diff
@@ -348,6 +377,32 @@ LY_ERR lyd_parse_lyb(const struct ly_ctx *ctx, const struct lysc_ext_instance *e
         struct lyd_node **first_p, struct ly_in *in, uint32_t parse_opts, uint32_t val_opts, uint32_t int_opts,
         struct ly_set *parsed, ly_bool *subtree_sibling, struct lyd_ctx **lydctx_p);
 
+#ifdef ENABLE_CBOR_SUPPORT
```
Again not necessary, the function can exist and fail. I would like to keep these macros to a minimum.

michalvasko commented on src/parser_json.c:
```diff
@@ -307,6 +307,7 @@ lydjson_get_snode(struct lyd_json_ctx *lydctx, ly_bool is_attr, const char *pref
             }
 
             /* unknown data node */
+            printf("checkpoint1-json\n");
```
Leftovers.

MeherRushi replied:
right sorry, will clear these

michalvasko commented on src/parser_json.c:
```diff
@@ -1067,7 +1068,7 @@ lydjson_parse_opaq(struct lyd_json_ctx *lydctx, const char *name, size_t name_le
             /* but first process children of the object in the array */
             do {
-                LY_CHECK_GOTO(ret = lydjson_subtree_r(lydctx, *node_p, lyd_node_child_p(*node_p), NULL), cleanup);
+                LY_CHECK_GOTO(ret = lydjson_subtree_r(lydctx, *node_p,                                                                                                                                                                                                                                                                                                                                          ...
```
This changes how JSON opaque nodes are stored, what is the reason?

MeherRushi replied:
I think this was some change made while trying to debug/understand code. Forgot to revert it back. Will correct it back

michalvasko commented on src/tree_data.h:
```diff
@@ -556,6 +557,7 @@ typedef enum {
                                  escaped when the anydata is printed in XML format. */
     LYD_ANYDATA_XML,        /**< Value is a string containing the serialized XML data. */
     LYD_ANYDATA_JSON,       /**< Value is a string containing the data modeled by YANG and encoded as I-JSON. */
+    LYD_ANYDATA_CBOR,       /**< Value is a string containing the data modeled by YANG and encoded as CBOR. */
```
`LYD_ANYDATA_LYB` was removed in the `devel` branch and I think for the same reason `LYD_ANYDATA_CBOR` is not required. Only objects are allowed in anydata, so it should always be possible to parse them into `LYD_ANYDATA_DATATREE`.

michalvasko reviewed (commented):
Generally looks fine but some improvements are possible and the coding style is somewhat different to ours.

MeherRushi commented:
@michalvasko Thank you for your reviews. Yup, we will correct as per the suggestions made and improve the coding style and make the PR to the devel branch. Sorry for the silly errors.

irfanHaslanded commented:
@MeherRushi I would like to know if this is still being worked on?

MeherRushi commented:
Yup, sorry for the delay. Will make the patches sometime soon

Commit:
fix: address PR review and migrate parser_cbor.c to current API

- Add FindCBOR.cmake module for proper CBOR dependency detection
- Remove debug code and leftover artifacts
- Revert parser_json.c to upstream (undo unintended changes)
- Fix CBOR struct layout in parser_internal.h, tree_data.c, printer_cbor.c
- Migrate parser_cbor.c to current upstream API:
  - Rewrite lydcbor_get_snode to use lys_find_child_node
  - Add lnode parameter to all LOGVAL calls
  - Fix LOG_LOCSET/LOG_LOCBACK to single-arg macros
  - Fix lyd_parser_create_term signature (add parent, value_size_bits)
  - Fix lyd_parser_create_meta signature (add lnode, value_size_bits)
  - Fix lyd_create_any to new 7-arg signature
  - Remove lyplg_ext_insert (function removed upstream)
  - Fix lyd_node_any access (value.tree → child)
  - Add tree_schema_internal.h include

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,129 @@
+ * @author MeherRushi <meherrrushi2@gmail.com>
+ * @brief CBOR data parser for libyang (abstraction over libcbor)
+ *
+ * Copyright (c) 2020 - 2023 CESNET, z.s.p.o.
```
Just `2026` should work here and in the other new files.

michalvasko commented on src/lcbor.h:
```diff
@@ -0,0 +1,81 @@
+ * @return LY_ERR value.
+ */
+LY_ERR
+lycbor_ctx_new(const struct ly_ctx *ctx, struct ly_in *in, struct lycbor_ctx **cborctx_p);
```
Functions declarations do not have a newline after the return value.

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @param[in] cborctx CBOR context to free.
+ */
+void lycbor_ctx_free(struct lycbor_ctx *cborctx)
```
Functions definitions use a newline after the return value.

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,129 @@
+ * @brief Free CBOR context.
+ *
+ * @param[in] cborctx CBOR context to free.
+ */
```
Non-static functions have their doxygen comment exclusively with their declaration.

michalvasko commented on src/lcbor.c:
```diff
@@ -0,0 +1,129 @@
+LY_ERR
+lycbor_ctx_new(const struct ly_ctx *ctx, struct ly_in *in, struct lycbor_ctx **cborctx_p)
+{
+    /* TODO : Need to restructure error handling here */
```
Fix the TODO.

michalvasko commented on src/parser_cbor.c:
```diff
@@ -0,0 +1,1882 @@
+
+    if (!item)
+    {
+        return LY_EINVAL;
```
Either use an `assert()` or add a proper error message, the return value on its own is of little use.

michalvasko commented on src/parser_cbor.c:
```diff
@@ -0,0 +1,1882 @@
+    case CBOR_TYPE_UINT: {
+        uint64_t val = cbor_get_int(item);
+        int len = snprintf(NULL, 0, "%" PRIu64, val);
+        if (len < 0) {
+            return LY_ESYS;
+        }
```
Redundant, the function cannot fail.

michalvasko commented on src/parser_data.h:
```diff
@@ -40,6 +40,12 @@ struct ly_in;
+ * - CBOR
+ *
+ *   The reference documentation would be `Encoding of Data Modeled with YANG in the Concise Binary Object 
+ *   Representation (CBOR)` : [RFC 9254](https://datatracker.ietf.org/doc/html/rfc9254).  < $TODO$ Look at the edge cases of 
+ *   RPCs, actions and Notifications and maybe like json only - where we make a proprietary representation>
```
What about this TODO?

michalvasko commented on src/printer_cbor.c:
```diff
@@ -0,0 +1,1308 @@
+                return NULL;
+            }
+            return mod->name;
+        default:
```
Prefer not using `default` so that adding to the values is detected by the compiler by a warning.

michalvasko commented on src/printer_data.c:
```diff
@@ -38,6 +38,11 @@ lyd_print_(struct ly_out *out, const struct lyd_node *root, LYD_FORMAT format, u
     case LYD_LYB:
         ret = lyb_print_data(out, root, options);
         break;
+#ifdef ENABLE_CBOR_SUPPORT
```
Proper format-unsupported error would be better.

michalvasko commented on src/tree_data.c:
```diff
@@ -131,6 +134,11 @@ lyd_parse(const struct ly_ctx *ctx, struct lyd_node *parent, struct lyd_node **f
     case LYD_LYB:
         r = lyd_parse_lyb(ctx, parent, first_p, in, parse_opts, val_opts, int_opts, &parsed, &lydctx);
         break;
+#ifdef ENABLE_CBOR_SUPPORT
```
Handle the case, with error unsupported format if CBOR not supported.

michalvasko commented on src/tree_data.c:
```diff
@@ -460,6 +484,11 @@ lyd_parse_op(const struct ly_ctx *ctx, struct lyd_node *parent, struct ly_in *in
     case LYD_LYB:
         rc = lyd_parse_lyb(ctx, parent, &first, in, parse_options, val_opts, int_opts, &parsed, &lydctx);
         break;
+#ifdef ENABLE_CBOR_SUPPORT
```
Handle the case, with error unsupported format if CBOR not supported.

michalvasko commented on src/tree_data.h:
```diff
@@ -513,9 +513,26 @@ typedef enum {
+    LYD_ANYDATA_JSON,       /**< Value is a string containing the data modeled by YANG and encoded as I-JSON. */
+    LYD_ANYDATA_LYB         /**< Value is a memory chunk with the serialized data tree in LYB format. */
+} LYD_ANYDATA_VALUETYPE;
+
```
Do not add this back, it is redundant in the current version.

michalvasko commented on CMakeLists.txt:
```diff
@@ -323,6 +324,16 @@ if(ENABLE_COVERAGE)
 
+find_package(CBOR)
+if(CBOR_FOUND)
+    set(CBOR_SUPPORT ON)
```
Seems like an unused variable.

MeherRushi commented:
@michalvasko Thanks for the review - will fix them soon

Commit:
fix: address PR review and formatting changes

    - Fix copyright years in all new files
    - Fix function declaration style and moving doxygen comments to the right place
    - Fixing other small formating errors and removing dead code

michalvasko commented:
You have not addressed all the comments and you also need to lint the code using `uncrustify` for the tests to pass.

MeherRushi commented:
> You have not addressed all the comments and you also need to lint the code using `uncrustify` for the tests to pass.

yes, yes , sorry working it out - one by one. will update once addressed.

michalvasko commented:
Finished in #2535 and merged.
